enhancements/image-references.md

```diff
@@ -0,0 +1,228 @@
+build-system-specific configuration to an absolute minimum.
+
+#### Story 4
+
+As an engineer involved in publishing an optional operator, I would like to
+configure semantically equivalent image pull specifications once, in order to not
+need to configure each build system independently.
+
+### Implementation Details/Notes/Constraints
+
+The core problems that must be solved in the implementation of this proposal have
+already been handled in the workflow used in `oc adm release new`. When implementing
+improvements to the `operator-sdk generate bundle` process we will simply need to
+create a shared library for the two tools to use. While the shape of the output
+is slightly different and some of the semantics about how the output should be 
+formatted are dissmilar, the core image reference rewriting is identical and the
+process of building a `FROM scratch` image layer is also identical.
+
+Today, some prior art exists in the OSBS workflow for building operator bundles.
+As we improve the Operator SDK tooling to create a straghtforward process for
+creating bundle images, we must make sure a seamless migration is possible.
+
+### Risks and Mitigations
+
+It will be critical that the design be vetted by all of the concerned parties, from
+operator manifest authors to CI system authors and productized pipeline authors to
+ensure that the UX is appropriate in all cases. Furthermore, the largest risk in the
+implementation here is not prioritizing a clean migration pathway for all current
+users who create bundle images, which would lead to further fragmentation of the
+ecosystem, which is directly opposed to the goal of this enhancement.
+
+## Design Details
+
+The definition of a minimally-viable operator bundle image will be changed to
+ensure that all image references in the contained manifests have been resolved 
+to a digest and had the pull specifications rewritten to refer to those digests.
+
+The only acceptable process for creating an operator bundle will be to run the
+`operator-sdk generate bundle` CLI, providing the manifests, metadata and image
+sources as input to the creation process.
+
+### Proposed UX
+
+Operator manifest authors write a manifest that refers to images using some
+opaque string [ex](https://github.com/openshift/machine-config-operator/blob/master/install/0000_80_machine-config-operator_04_deployment.yaml#L22),
+and provide an `image-references` file alongside their manifests that declares
+which strings inside of their manifest are referring to pull specifications of
+images and names each occurence [ex](https://github.com/openshift/machine-config-operator/blob/98d9ba6841eb4811ed6f4d4de7016ea83c131c54/install/image-references#L7-L10).
+
+The `metadata/image-references` file holds data in the following format, mapping
+common names of container imags to their string placeholders in the manifests:
+
+```yaml
+imageReferences:
+- name: common-name
+  placeholder: registry.svc.ci.openshift.org/openshift:image
```

Comment:
Quick bikeshed on `placeholder`, what about aligning somewhat with `sed` terminology, like:

 - `substitute`
 - `pattern`
 
and do we think we will only ever allow full-text matching?

Reply:
`substitute` also sounds good, I don't feel strongly on this.

Reply:
We don't want to use [an `ImageStream`][1] like [OpenShift release images do][2]?

[1]: https://github.com/openshift/machine-config-operator/blob/915ca7a11fead41d75ffa0ed7e310f7284645a13/install/image-references
[2]: https://github.com/openshift/cluster-version-operator/blob/62ff8e3e51cd4bfe6cd3b8cb6400a28ada1b183d/docs/dev/operators.md#how-do-i-ensure-the-right-images-get-used-by-my-manifests

Reply:
That was proposed at the start, but an OpenShift-agnostic approach was suggested by the maintainers. I think a translation library from an ImageStream to these more targeted formats is simple to write, so there is no harm in using them. As factored today, a list of `struct`s allows us room in the future to expand the API surface in a simple but backwards compatible manner, too.

Reply:
Talk about that choice in an *Alternatives* section, so folks curious about the divergence don't have to dig it out of PR discussion?